File: python_server/modules/verifier.py
"""
AI Travel Butler - Plan Verification
"""

from typing import List, Dict, Any


def verify_plan(items: List[Dict], mcp_trace: List[str]) -> List[str]:
    """Verify itinerary and return warnings."""
    warnings = []

    if not items or len(items) == 0:
        warnings.append("未生成任何地图标注点，请重试或切换模型。")
        mcp_trace.append("agent:verifier:warnings:1:no_locations")
        return warnings

    by_day = {}
    for item in items:
        day = item.get("day", 1)
        if day not in by_day:
            by_day[day] = []
        by_day[day].append(item)

    import math

    for day, day_items in by_day.items():
        if len(day_items) < 3:
            warnings.append(f"Day {day} 行程点位少于 3 个，建议补充早餐/晚间活动/交通节点。")

        sequences = sorted([int(x.get("sequence", 0)) for x in day_items])
        for i in range(1, len(sequences)):
            if sequences[i] == sequences[i - 1]:
                warnings.append(f"Day {day} 存在重复 sequence={sequences[i]}，可能导致顺序冲突。")
                break

        for i in range(len(day_items) - 1):
            current = day_items[i]
            next_item = day_items[i + 1]

            try:
                curr_lat = float(current.get("lat", 0))
                curr_lng = float(current.get("lng", 0))
                next_lat = float(next_item.get("lat", 0))
                next_lng = float(next_item.get("lng", 0))

                if curr_lat and curr_lng and next_lat and next_lng:
                    lat_diff = abs(curr_lat - next_lat)
                    lng_diff = abs(curr_lng - next_lng)
                    if lat_diff > 0.5 and lng_diff > 0.5:
                        warnings.append(f"Day {day} 中 {current.get('name')} 和 {next_item.get('name')} 之间距离过远，可能不连贯。")
                        mcp_trace.append(f"agent:verifier:geo_gap_warning:day_{day}")
            except (ValueError, TypeError):
                pass

    unique_cities = set(item.get("city") for item in items if item.get("city"))
    if len(unique_cities) > 1:
        warnings.append(f"行程包含了多个城市: {', '.join(unique_cities)}，不符合目的地要求。")
        mcp_trace.append("agent:verifier:multi_city_warning")

    if warnings:
        mcp_trace.append(f"agent:verifier:warnings:{len(warnings)}")
    else:
        mcp_trace.append("agent:verifier:pass")

    return warnings

File: server/modules/verifier.mjs
/**
 * AI Travel Butler - Plan Validator
 * 行程验证和校验工具
 */

export function verifyPlan(items, mcpTrace) {
  const warnings = [];
  if (!items || items.length === 0) {
    warnings.push('未生成任何地图标注点，请重试或切换模型。');
    mcpTrace.push('agent:verifier:warnings:1:no_locations');
    return warnings;
  }
  const byDay = new Map();
  for (const item of items) {
    if (!byDay.has(item.day)) byDay.set(item.day, []);
    byDay.get(item.day).push(item);
  }

  for (const [day, dayItems] of byDay.entries()) {
    if (dayItems.length < 3) {
      warnings.push(`Day ${day} 行程点位少于 3 个，建议补充早餐/晚间活动/交通节点。`);
    }
    const seq = dayItems.map((x) => Number(x.sequence || 0)).sort((a, b) => a - b);
    for (let i = 1; i < seq.length; i += 1) {
      if (seq[i] === seq[i - 1]) {
        warnings.push(`Day ${day} 存在重复 sequence=${seq[i]}，可能导致顺序冲突。`);
        break;
      }
    }
    
    // 增强: 检查地理连贯性
    for (let i = 0; i < dayItems.length - 1; i++) {
      const current = dayItems[i];
      const next = dayItems[i + 1];
      
      // 解析经纬度字符串为数值
      const currLat = parseFloat(current.lat);
      const currLng = parseFloat(current.lng);
      const nextLat = parseFloat(next.lat);
      const nextLng = parseFloat(next.lng);
      
      if (!isNaN(currLat) && !isNaN(currLng) && !isNaN(nextLat) && !isNaN(nextLng)) {
        // 计算两点之间的欧氏距离（简化）- 如果相隔太远可能地理不连贯
        const latDiff = Math.abs(currLat - nextLat);
        const lngDiff = Math.abs(currLng - nextLng);
        if (latDiff > 0.5 && lngDiff > 0.5) { // 距离阈值约为50公里（粗略估计）
          warnings.push(`Day ${day} 中 ${current.name} 和 ${next.name} 之间距离过远，可能不连贯。`);
          mcpTrace.push(`agent:verifier:geo_gap_warning:day_${day}`);
        }
      }
    }
    
    // 增强: 验证时间安排合理性
    const sortedByTime = dayItems.sort((a, b) => {
      // 简单比较时间字符串 HH:MM 格式
      return a.time.localeCompare(b.time);
    });
    let prevTime = null;
    for (const item of sortedByTime) {
      if (prevTime && item.time && prevTime === item.time) {
        warnings.push(`Day ${day} 存在时间冲突：${item.name} 和 ${byDay.get(day)[sortedByTime.indexOf(item)].name} 安排在同一时间。`);
      }
      prevTime = item.time;
    }
    
    // 增强: 颺算约束检查（基于估算）
    if (dayItems.some(item => item.estimated_cost && parseInt(item.estimated_cost) > 500)) {
      warnings.push(`Day ${day} 包含较高消费景点，可能超出低预算预期。`);
      mcpTrace.push('agent:verifier:budget_warning');
    }
  }

  // 检查是否所有地点都在同一个城市，如果不一致可能行程不连贯
  const uniqueCities = [...new Set(items.map(item => item.city))];
  if (uniqueCities.length > 1) {
    warnings.push(`行程包含了多个城市: ${uniqueCities.join(', ')}，不符合目的地要求。`);
    mcpTrace.push('agent:verifier:multi_city_warning');
  }

  if (warnings.length === 0) {
    mcpTrace.push('agent:verifier:pass');
  } else {
    mcpTrace.push(`agent:verifier:warnings:${warnings.length}`);
  }

  return warnings;
}